tests/infra_test.go

```diff
@@ -54,6 +55,121 @@ var _ = Describe("Infrastructure", func() {
+			Body(vmi).
+			Do().Get()
+		Expect(err).ToNot(HaveOccurred(), "Should create VMI")
+
+		By("Waiting until the VM is ready")
+		return tests.WaitForSuccessfulVMIStart(obj)
+	}
+
+	Describe("Prometheus scraped metrics", func() {
+
+		/*
+			This test is querying the metrics from Prometheus *after* they were
+			scraped and processed by the different components on the way.
+		*/
+
+		tests.BeforeAll(func() {
+			onOCP, err := clusterutil.IsOnOpenShift(virtClient)
+			Expect(err).ToNot(HaveOccurred(), "failed to detect cluster type")
+
+			if !onOCP {
+				Skip("test is verifying integration with OCP's cluster monitoring stack")
+			}
+		})
+
+		It("should find VMI namespace on namespace label of the metric", func() {
+
+			/*
+				This test is required because in cases of misconfigurations on
+				monitoring objects (such for the ServiceMonitor), our rules will
+				still be picked up by the monitoring-operator, but Prometheus
+				will fail to load it.
+			*/
+
+			By("creating a VMI in a user defined namespace")
+			vmi := tests.NewRandomVMIWithEphemeralDisk(
+				tests.ContainerDiskFor(tests.ContainerDiskAlpine))
+			startVMI(vmi)
+
+			By("finding virt-operator pod")
+			ops, err := virtClient.CoreV1().Pods("kubevirt").
+				List(metav1.ListOptions{LabelSelector: "kubevirt.io=virt-operator"})
+			Expect(err).ToNot(HaveOccurred(), "failed to list virt-operators")
+			Expect(ops.Size).ToNot(Equal(0), "no virt-operators found")
+			op := ops.Items[0]
+			Expect(op).ToNot(BeNil(), "virt-operator pod should not be nil")
+
+			By("finding Prometheus endpoint")
+			ep, err := virtClient.CoreV1().Endpoints("openshift-monitoring").
+				Get("prometheus-k8s", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred(), "failed to retrieve Prometheus endpoint")
+			promIP := ep.Subsets[0].Addresses[0].IP
+			Expect(promIP).ToNot(Equal(0), "could not get Prometheus IP from endpoint")
+			var promPort int32
+			for _, port := range ep.Subsets[0].Ports {
+				if port.Name == "web" {
+					promPort = port.Port
+				}
+			}
+			Expect(promPort).ToNot(Equal(0), "could not get Prometheus port from endpoint")
+
+			// We need a token from a service account that can view all namespaces in the cluster
+			By("extracting virt-operator sa token")
+			token, _, err := tests.ExecuteCommandOnPodV2(virtClient,
+				&op,
+				"virt-operator",
+				[]string{
+					"cat",
+					"/var/run/secrets/kubernetes.io/serviceaccount/token",
+				})
+			Expect(err).ToNot(HaveOccurred(), "failed executing command on virt-operator")
+			Expect(token).ToNot(BeEmpty(), "virt-operator sa token returned empty")
+
+			By("querying Prometheus API endpoint for a VMI exported metric")
+			stdout, _, err := tests.ExecuteCommandOnPodV2(virtClient,
+				&op,
+				"virt-operator",
+				[]string{
+					"curl",
+					"-L",
+					"-k",
+					fmt.Sprintf("https://%s:%d/api/v1/query", promIP, promPort),
+					"-H",
+					fmt.Sprintf("Authorization: Bearer %s", token),
```

Comment:
you don't actually have to extract the bearer token from the pod do you? Because the right token is already in the pod you're executing this command within. 

What would happen if you replaced this line with something line the following instead?

"Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"

Reply:
I think that `kubectl exec` does not support subshells or at least I couldn't make it escape the subshell somehow and ensure that its being interpreted inside the container.

Reply:
It is not that straight forward as it seems at first... I've tried this approach before and the problem is that the execution of a sub-shell does not seem to work as I expected after processed by `exec`.

Even if I can get it to work, I think that it will have a negative impact on the readability of this test.